File: projects/func/OC/oneOC_mac/Model/PacketH.m
//
//  PacketH.m
//  oneOC
//
//  Created by Jz D on 2020/4/3.
//  Copyright © 2020 Jz D. All rights reserved.
//

#import "PacketH.h"

NSString * const MTPacketKeyData = @"data";
NSString * const MTPacketKeyType = @"type";
NSString * const MTPacketKeyAction = @"action";

@implementation PacketH

#pragma mark -
#pragma mark Initialization
- (instancetype)initWithData:(id)data type: (PacketType)type action: (PacketAction)action {
    self = [super init];
 
    if (self) {
        self.data = data;
        self.type = type;
        self.action = action;
    }
 
    return self;
}
 
#pragma mark -
#pragma mark NSCoding Protocol
- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeObject:self.data forKey:MTPacketKeyData];
    [coder encodeInteger:self.type forKey:MTPacketKeyType];
    [coder encodeInteger:self.action forKey:MTPacketKeyAction];
}




- (instancetype)initWithCoder:(NSCoder *)coder{
    self = [super init];
    if (self) {
        _data = [coder decodeObjectForKey: MTPacketKeyData];
        _type = [coder decodeIntegerForKey: MTPacketKeyType];
        _action = [coder decodeIntegerForKey: MTPacketKeyAction];
    }
 
    return self;
}


+ (BOOL)supportsSecureCoding {
   return YES;
}




@end

File: projects/func/socketG/socketG/Model/PacketH.swift
//
//  PacketH.swift
//  socketD
//
//  Created by Jz D on 2020/4/16.
//  Copyright © 2020 Jz D. All rights reserved.
//

import Foundation


enum PacketType: Int{
    case unknown = -1, didAddDisc, startNewTask
}


enum PacketAction: Int{
    case unknown = -1
    case go = 0
}




struct PacketKey {
    static let data = "data"
    static let type = "type"
    static let action = "action"
}


class PacketH: NSObject{


    let data: Any
    let type: PacketType
    let action: PacketAction
    

    
    init(info d: Any, type t: PacketType, action a: PacketAction){
        data = d
        type = t
        action = a
        super.init()
    }
    
    
    required init?(coder: NSCoder) {
        data = coder.decodeObject(forKey: PacketKey.data) ?? [:]
        type = PacketType(rawValue: coder.decodeInteger(forKey: PacketKey.type)) ?? PacketType.unknown
        action = PacketAction(rawValue: coder.decodeInteger(forKey: PacketKey.action)) ?? PacketAction.unknown
    }
    
}


extension PacketH: NSCoding, NSSecureCoding{
    
    
    
    func encode(with coder: NSCoder) {
        coder.encode(data, forKey: PacketKey.data)
        coder.encode(type.rawValue, forKey: PacketKey.type)
        coder.encode(action.rawValue, forKey: PacketKey.action)
    }
    
    
    
    
    
    
    static var supportsSecureCoding: Bool {
        true
    }
    
    
    
    
    
}
